### PR TITLE
Gnu apl

### PR DESCRIPTION
**Pull request by casouri**
#### Description
Add port `gnu-apl` and select group `apl`.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.1 19B88
Xcode 11.3 11C29

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: lang/gnu-apl/Portfile [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+PortSystem          1.0
+PortGroup           select 1.0
+
+name                gnu-apl
+version             1.8
+categories          lang
+platforms           darwin
+license             GPL
+maintainers         {@casouri gmail.com:casouri}
```
Comment (pmetzger) [inline lang/gnu-apl/Portfile]: You probably want to add `openmaintainer` to the end here unless you have good reason not to, as in:

```
maintainers         {@casouri gmail.com:casouri} openmaintainer
```

File: lang/gnu-apl/Portfile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+description         GNU APL is a free interpreter for the programming language APL.
+
+long_description    GNU APL is a free interpreter for the programming language APL. \
+                    The APL interpreter is an (almost) complete implementation of ISO \
+                    standard 13751 aka. Programming Language APL, Extended.
+
+homepage            https://www.gnu.org/software/apl/
+master_sites        https://ftp.gnu.org/gnu/apl/ \
+                    https://ftpmirror.gnu.org/apl/
```
Comment (pmetzger) [inline lang/gnu-apl/Portfile]: Won't `master_sites gnu:apl` work?

File: lang/gnu-apl/files/gnu-apl [hunk 3]
```diff
@@ -0,0 +1 @@
+bin/gnu-apl
```
Comment (pmetzger) [inline lang/gnu-apl/files/gnu-apl]: What is the function of this file?

File: sysutils/apl_select/Portfile [hunk 4]
```diff
@@ -0,0 +1,28 @@
+PortSystem 1.0
```
Comment (pmetzger) [inline sysutils/apl_select/Portfile]: This file should start with the usual `#` prefixed modeline. It's also got some indentation problems.

File: sysutils/apl_select/Portfile [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+PortGroup select 1.0
+
+name                    apl_select
+version                 1.0
+categories              sysutils
+platforms               darwin
+license                 BSD
+maintainers             {@casouri gmail.com:casouri}
```
Comment (pmetzger) [inline sysutils/apl_select/Portfile]: Again, please add `openmaintainer`

File: sysutils/apl_select/Portfile [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+supported_archs         noarch
+description             Common files for selecting default apl version
+long_description        apl_select installs files that allow 'port select' to \
+			switch the default version of apl. \
+			It symlinks the standard apl executables \
+			in the MacPorts prefix to the selected version.
```
Comment (pmetzger) [inline sysutils/apl_select/Portfile]: Indentation here is a bit messed up.

File: sysutils/apl_select/files/base [hunk 7]
```diff
@@ -0,0 +1 @@
+bin/apl
```
Comment (pmetzger) [inline sysutils/apl_select/files/base]: What is the function of this file?

File: sysutils/apl_select/files/none [hunk 8]
```diff
@@ -0,0 +1 @@
+-
```
Comment (pmetzger) [inline sysutils/apl_select/files/none]: What is the function of this file?

File: lang/gnu-apl/Portfile (hunk 1 above)
Comment (casouri) [reply lang/gnu-apl/Portfile]: Done.

File: lang/gnu-apl/Portfile (hunk 2 above)
Comment (casouri) [reply lang/gnu-apl/Portfile]: Done.

File: lang/gnu-apl/files/gnu-apl (hunk 3 above)
Comment (casouri) [reply lang/gnu-apl/files/gnu-apl]: Do you mean I should add comments to the file?

File: sysutils/apl_select/Portfile (hunk 4 above)
Comment (casouri) [reply sysutils/apl_select/Portfile]: Done.

File: sysutils/apl_select/Portfile (hunk 6 above)
Comment (casouri) [reply sysutils/apl_select/Portfile]: Ah, don't know why tab muggled in.

**Comment by casouri** [comment]
Thanks, here is the update. I'll squash them when there is no more to change.

File: lang/gnu-apl/files/gnu-apl (hunk 3 above)
Comment (casouri) [reply lang/gnu-apl/files/gnu-apl]: Like `bin/gnu-apl # this is the executable for APL interpreter`?

File: lang/gnu-apl/Portfile [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+maintainers         {@casouri gmail.com:casouri} openmaintainer
+description         GNU APL is a free interpreter for the programming language APL.
+
+long_description    GNU APL is a free interpreter for the programming language APL. \
+                    The APL interpreter is an (almost) complete implementation of ISO \
+                    standard 13751 aka. Programming Language APL, Extended.
+
+homepage            https://www.gnu.org/software/apl/
+master_sites        gnu:apl
+distfiles           apl-1.8.tar.gz
```
Comment (mf2k) [inline lang/gnu-apl/Portfile]: Don't set `distfiles` directly. Set the `distname` instead. Also don't hardcode the `version`.

Comment (casouri) [reply lang/gnu-apl/Portfile]: Done

**Comment by casouri** [comment]
update3 is some minor change to the notes.

File: sysutils/apl_select/Portfile [hunk 10]
```diff
@@ -25,4 +27,4 @@ destroot {
     select::install apl ${filespath}/none
 }
 
-livecheck.type	none
+livecheck.type          none
```
Comment (pmetzger) [inline sysutils/apl_select/Portfile]: BTW, having a livecheck is a good idea in general. They're easy to do, so you should probably have one. Check another package (like gnutar say) that has a `gnu:` dist location in it.

File: lang/gnu-apl/files/gnu-apl (hunk 3 above)
Comment (pmetzger) [reply lang/gnu-apl/files/gnu-apl]: But it isn't. It's just a one line file named "gnu-apl" with the words "bin/gnu-apl" in it.

File: sysutils/apl_select/Portfile (hunk 10 above)
Comment (casouri) [reply sysutils/apl_select/Portfile]: Thanks, done for gnu-apl.

File: lang/gnu-apl/files/gnu-apl (hunk 3 above)
Comment (casouri) [reply lang/gnu-apl/files/gnu-apl]: Ah, you are asking the function of files/gnu-apl. I thought you are asking the function of /bin/gnu-apl.

This file works with `apl_select`, so `port select --set apl gnu-apl` can symlink `gnu-apl` to `apl`. Also `apl_select/files/base` and `apl_select/files/none`. The files in `base` lays out the destination files and `gnu-apl` lists the source. And `-` in `none` means don't symlink. 

That's how I understand it and not guaranteed to be true. There is no official documentation at all and I asked on IRC and copied from other portfiles.

File: lang/gnu-apl/Portfile [hunk 11]
```diff
@@ -28,6 +28,12 @@ checksums           rmd160  e277ae100c8eccae32980b606401b04bfe65b0b1 \
                     sha256  144f4c858a0d430ce8f28be90a35920dd8e0951e56976cb80b55053fa0d8bbcb \
                     size    3565018
 
+<<<<<<< Updated upstream
+=======
+livecheck.type      regex
+livecheck.regex     apl-(\[0-9.\]+)${extract.suffix}
+
+>>>>>>> Stashed changes
```
Comment (pmetzger) [inline lang/gnu-apl/Portfile]: You broke this.

Comment (casouri) [reply lang/gnu-apl/Portfile]: Sorry, apparently I'm not very good at git.

File: sysutils/apl_select/files/base (hunk 7 above)
Comment (pmetzger) [reply sysutils/apl_select/files/base]: Note that this file lacks a terminating carriage return...

File: lang/gnu-apl/files/gnu-apl (hunk 3 above)
Comment (pmetzger) [reply lang/gnu-apl/files/gnu-apl]: Does it seem to work correctly?

Comment (casouri) [reply lang/gnu-apl/files/gnu-apl]: Yeah. Selecting `gnu-apl` and `none` work.

File: sysutils/apl_select/files/base (hunk 7 above)
Comment (casouri) [reply sysutils/apl_select/files/base]: Added.

**Comment by casouri** [comment]
Oops, I click "re-request review" and something changed. I hope it doesn't harm.

**Comment by pmetzger** [comment]
@mf2k Are you happy with this at this point? If so, you can squash and merge, I won't be able to get to it for a while.

**Comment by casouri** [comment]
Any news?

**Comment by mf2k** [comment]
The commits should have been squashed first.

**Comment by casouri** [comment]
Hmmm, can you undo the merge? Maybe a force push before anyone noticed.

**Comment by casouri** [comment]
What should I do now? Should I squash my fork?

**Comment by ryandesign** [comment]
It's too late to do anything; the commits have landed on master. We do not allow force pushing to master.

In the future, let's please make sure we squash intermediate commits down to the smallest reasonable number of logically separate commits, with commit messages that conform to our guidelines.

The PR submitter does not necessarily have to be the one to do this. The person merging the PR can squash all commits into one by using the "Squash and merge" button instead of the "Rebase and merge" button; the commit message can and should be edited in the web interface in that case. If wishing to squash down to more than one commit, that can be done somewhat more laboriously on the command line.

**Pull request merged**